File: wimg-ios/wimg/Models/Transaction.swift
import Foundation

struct Transaction: Codable, Identifiable {
    let id: String
    let date: String
    let description: String
    let amount: Double
    let currency: String
    let category: Int

    var isExpense: Bool { amount < 0 }
    var isIncome: Bool { amount > 0 }
    var absAmount: Double { abs(amount) }

    var dateFormatted: String {
        // date comes as "YYYY-MM-DD"
        let parts = date.split(separator: "-")
        guard parts.count == 3 else { return date }
        return "\(parts[2]).\(parts[1]).\(parts[0])"
    }

    var parsedDate: Date? {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.date(from: date)
    }
}

struct ImportResult: Codable {
    let total_rows: Int
    let imported: Int
    let skipped_duplicates: Int
    let errors: Int
    let format: String
    let categorized: Int
}

File: wimg-android/app/src/main/java/com/wimg/app/models/Transaction.kt
package com.wimg.app.models

import kotlinx.serialization.Serializable
import kotlin.math.abs

@Serializable
data class Transaction(
    val id: String,
    val date: String,
    val description: String,
    val amount: Double,
    val currency: String = "EUR",
    val category: Int = 0,
    val account: String? = null,
    val excluded: Int? = null,
) {
    val isExpense: Boolean get() = amount < 0
    val isIncome: Boolean get() = amount > 0
    val absAmount: Double get() = abs(amount)
    val isExcluded: Boolean get() = (excluded ?: 0) != 0
}

@Serializable
data class ImportResult(
    val total_rows: Int,
    val imported: Int,
    val skipped_duplicates: Int,
    val errors: Int,
    val format: String,
    val categorized: Int,
)

@Serializable
data class ParseResult(
    val format: String,
    val total_rows: Int,
    val transactions: List<Transaction>,
)

@Serializable
data class UndoResult(
    val op: String,
    val table: String,
    val row_id: String,
    val column: String? = null,
)
